README
python-edl
==========

A python EDL (Edit Decision List) parsing library, based on the ruby EDL
library by Julik Tarkhanov http://guerilla-di.org/edl/. Still a work in
progress so collaboration welcome.

Uses pytimecode.py to handle the Timecode mathematics and should run on
Python 2.7+

https://github.com/simonh10/python-edl

Usage::

    from edl.edl import Parser
    parser=Parser('23.98')
    with open('file.edl') as f:
        edl=parser.parse(f)
        for event in edl.events:
            print "Event Number:"+str(event.num)
            print "Source file:"+str(event.source_file)
            print "Clip Name:"+str(event.clip_name)

Accepted framerate values ['60', '59.94', '50', '30', '29.97', '25', '24',
'23.98'] whole number frame rates are more tested than others, but the accuracy
relies heavily on accuracy of the pytimecode library.

Documentation for pytimecode can be found
[here](https://code.google.com/p/pytimecode/).


(The MIT License)

Copyright © 2013 Simon Hargreaves <simon@simon-hargreaves.com>

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the ‘Software’), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
of the Software, and to permit persons to whom the Software is furnished to do
so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED ‘AS IS’, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


CHANGES
v0.1, 11/06/2013 -- Initial release


tests/test_edl.py
#!/usr/bin/python
"""
Module EDL
unit test suite
"""

from edl import edl
import unittest


class TestEDLFunctions(unittest.TestCase):
    """tests the edl functions
    """

    def test_24fps(self):
        p = edl.Parser('24')
        with open('../tests/test-edls/test_24.edl') as f:
            s = p.parse(f)

        self.assertEqual(s.events[0].clip_name, 'clip 1',
                         'Failed clip name test')
        self.assertEqual(s.events[0].src_length(), 1440,
                         'Wrong source frame length')
        self.assertEqual(s.events[0].rec_length(), 1440,
                         'Wrong record frame length')
        self.assertEqual(s.events[0].src_start_tc.frames, 86400,
                         'Wrong source start timecode')
        self.assertEqual(s.events[0].src_end_tc.frames, 87840,
                         'Wrong source end timecode')
        self.assertEqual(s.events[0].rec_start_tc.frames, 0,
                         'Wrong record start timecode')
        self.assertEqual(s.events[0].rec_end_tc.frames, 1440,
                         'Wrong record end timecode')
        self.assertEqual(s.events[1].clip_name, 'clip #2',
                         'Failed clip name test char 2')
        self.assertEqual(s.events[2].clip_name, 'clip -3',
                         'Failed clip name test char 3')
        self.assertEqual(s.events[3].clip_name, 'clip $4',
                         'Failed clip name test char 4')
        self.assertEqual(s.events[4].clip_name, 'clip &5',
                         'Failed clip name test char 5')
        self.assertEqual(s.events[5].src_start_tc.frames, 697,
                         "Wrong Source start complex event")
        self.assertEqual(s.events[5].src_end_tc.frames, 697,
                         "Wrong Source end complex event")
        self.assertEqual(s.events[5].rec_start_tc.frames, 2857,
                         "Wrong Source start complex event")
        self.assertEqual(s.events[5].rec_end_tc.frames, 2857,
                         "Wrong Source end complex event")

    def test_pal(self):
        p = edl.Parser('25')
        with open('../tests/test-edls/test_25.edl') as f:
            s = p.parse(f)

    def test_ntsc(self):
        p = edl.Parser('29.97')
        with open('../tests/test-edls/test_2997NDF.edl') as f:
            s = p.parse(f)

    def test_2398fps(self):
        p = edl.Parser('23.98')
        with open('../tests/test-edls/test_2398.edl') as f:
            s = p.parse(f)


tests/test_list.py
# -*- coding: utf-8 -*-

import unittest
import edl


class ListTestCase(unittest.TestCase):
    """tests the edl.edl.List class
    """

    def setup(self):
        """set up the tests
        """
        pass

    def test_repr_is_matching_Python_standard(self):
        """testing if the List.__repr__ method output is matching Python
        standard of <edl.edl.List object at 0x??????> format
        """
        self.fail('test is not implemented yet')

    def testing_to_edl_method_will_output_the_standard_edl_case1(self):
        """testing if to_string will output the EDL as string
        """
        p = edl.Parser('24')
        with open('../tests/test_data/test_24.edl') as f:
            s = p.parse(f)

        with open('../tests/test_data/test_24.edl') as f:
            expected_edl = f.readlines()

        self.assertEqual(
            ''.join(expected_edl),
            s.to_string()
        )

    def testing_to_edl_method_will_output_the_standard_edl_case2(self):
        """testing if to_string will output the EDL as string
        """
        p = edl.Parser('24')
        with open('../tests/test_data/test.edl') as f:
            s = p.parse(f)

        with open('../tests/test_data/test.edl') as f:
            expected_edl = f.readlines()

        print s.to_string()

        self.assertEqual(
            ''.join(expected_edl),
            s.to_string()
        )

    def testing_to_edl_method_will_output_the_standard_edl_case3(self):
        """testing if to_string will output the EDL as string
        """
        p = edl.Parser('24')
        with open('../tests/test_data/test_50.edl') as f:
            s = p.parse(f)

        with open('../tests/test_data/test_50.edl') as f:
            expected_edl = f.readlines()

        print s.to_string()

        self.assertEqual(
            ''.join(expected_edl),
            s.to_string()
        )


tests/test_parser.py
#!/usr/bin/python
"""
Module EDL
unit test suite
"""

from edl import edl
import unittest


class ParserTestCase(unittest.TestCase):
    """tests the Parser
    """

    def test_24fps(self):
        p = edl.Parser('24')
        with open('../tests/test_data/test_24.edl') as f:
            s = p.parse(f)

        self.assertEqual(s.events[0].clip_name, 'clip 1',
                         'Failed clip name test')
        self.assertEqual(s.events[0].src_length(), 1440,
                         'Wrong source frame length')
        self.assertEqual(s.events[0].rec_length(), 1440,
                         'Wrong record frame length')
        self.assertEqual(s.events[0].src_start_tc.frames, 86400,
                         'Wrong source start timecode')
        self.assertEqual(s.events[0].src_end_tc.frames, 87840,
                         'Wrong source end timecode')
        self.assertEqual(s.events[0].rec_start_tc.frames, 0,
                         'Wrong record start timecode')
        self.assertEqual(s.events[0].rec_end_tc.frames, 1440,
                         'Wrong record end timecode')
        self.assertEqual(s.events[1].clip_name, 'clip #2',
                         'Failed clip name test char 2')
        self.assertEqual(s.events[2].clip_name, 'clip -3',
                         'Failed clip name test char 3')
        self.assertEqual(s.events[3].clip_name, 'clip $4',
                         'Failed clip name test char 4')
        self.assertEqual(s.events[4].clip_name, 'clip &5',
                         'Failed clip name test char 5')
        self.assertEqual(s.events[5].src_start_tc.frames, 697,
                         "Wrong Source start complex event")
        self.assertEqual(s.events[5].src_end_tc.frames, 697,
                         "Wrong Source end complex event")
        self.assertEqual(s.events[5].rec_start_tc.frames, 2857,
                         "Wrong Source start complex event")
        self.assertEqual(s.events[5].rec_end_tc.frames, 2857,
                         "Wrong Source end complex event")

    def test_pal(self):
        p = edl.Parser('25')
        with open('../tests/test_data/test_25.edl') as f:
            s = p.parse(f)

    def test_ntsc(self):
        p = edl.Parser('29.97')
        with open('../tests/test_data/test_2997NDF.edl') as f:
            s = p.parse(f)

    def test_2398fps(self):
        p = edl.Parser('23.98')
        with open('../tests/test_data/test_2398.edl') as f:
            s = p.parse(f)


edl/edl.py
#-*- coding: utf-8 -*-
"""
Module EDL
"""

import sys
import collections
import re
import pprint
import pytimecode


class List(object):
    """The EDL it self.

    Holds :class:`.Event` instances. It can be indexed to reach each of the
    :class:`.Event`\ s like::

      >>> l = List('25')
      >>> l.append(Event())
      >>> l[0]
      <edl.edl.Event object at 0x7fb630564490>

    :param str fps: The frame per second setting for for this EDL. Should be a
      string one of ['23.98', '24', '25', '29.97', '30', '50', '59.94', '60'].
      `fps` can not be skipped.
    """

    def __init__(self, fps):
        self.events = []
        self.fps = fps
        self.title = ''

    def __getitem__(self, i):
        """Returns each of the Events that this List holds.
        """
        return self.events[i]

    # def __repr__(self):
    #     rep = ["event(\n"]
    #     for e in self.events:
    #         rep.append(e.__repr__())
    #     rep.append(')')
    #     return ''.join(rep)

    def __len__(self):
        return len(self.events)

    def get_start(self):
        start_tc = None
        for e in self.events:
            if start_tc is None:
                start_tc = e.rec_start_tc
            else:
                if e.rec_start_tc.frames < start_tc.frames:
                    start_tc = e.rec_start_tc
        return start_tc

    def get_end(self):
        end_tc = None
        for e in self.events:
            if end_tc is None:
                end_tc = e.rec_end_tc
            else:
                if e.rec_end_tc.frames > end_tc.frames:
                    end_tc = e.rec_end_tc
        return end_tc

    def get_length(self):
        return self.get_end().frames - self.get_start().frames

    def append(self, evt):
        self.events.append(evt)

    def events(self):
        return self.events

    def without_transitions(self):
        raise NotImplementedError

    def renumbered(self):
        raise NotImplementedError

    def without_timewarps(self):
        raise NotImplementedError

    def without_generators(self):
        raise NotImplementedError

    def capture_list(self):
        raise NotImplementedError

    def from_zero(self):
        raise NotImplementedError

    def spliced(self):
        raise NotImplementedError

    def to_string(self):
        """The string output of the Events, this matches a standard EDL file
        format. Using List.to_string() should return the edl back to its
        original format.
        """
        # for each Event call their to_string() method and gather the output
        output_buffer = ['TITLE: %s' % self.title, '']
        for event in self.events:
            output_buffer.append(event.to_string())
            # output_buffer.append('')
        return '\n'.join(output_buffer)


class Matcher(object):
    """No documentation for this class yet.
    """

    def __init__(self, with_regex):
        self.regex = with_regex

    def matches(self, line):
        return re.match(self.regex, line)

    def apply(self, stack, line):
        sys.stderr.write("Skipping:" + line)


class TitleMatcher(Matcher):
    """Matches the EDL Title attribute
    """
    def __init__(self):
        Matcher.__init__(self, 'TITLE: (.+)')

    def apply(self, stack, line):
        m = re.search(self.regex, line)
        try:
            stack.title = m.group(1).strip()
        except (IndexError, AttributeError):
            pass


class CommentMatcher(Matcher):
    """No documentation for this class yet.
    """

    def __init__(self):
        Matcher.__init__(self, '\*\s*(.+)')

    def apply(self, stack, line):
        #print line
        m = re.search(self.regex, line)
        if m:
            if len(stack) > 0:
                stack[-1].comments.append("* " + m.group(1))
                mo = re.search('\*\s+FROM\s+CLIP\s+NAME:\s+(.+)', line)
                if mo:
                    stack[-1].clip_name = mo.group(1).strip()


class FallbackMatcher(Matcher):
    """No documentation for this class yet.
    """

    def __init__(self):
        Matcher.__init__(self, '/^(\w)(.+)/')

    def apply(self, stack, line):
        pass


class NameMatcher(Matcher):
    """No documentation for this class yet.
    """

    def __init__(self):
        # TODO: shouldn't it be '\*\s*FROM\s+CLIP\s+NAME:(\s+)(.+)' as above,
        #       add a test for this
        Matcher.__init__(self, '\*\s*FROM CLIP NAME:(\s+)(.+)')

    def apply(self, stack, line):
        m = re.search(self.regex, line)
        #print line
        if len(stack) > 0:
            if m:
                stack[-1].clip_name = m.group(2).strip()


class SourceMatcher(Matcher):
    """No documentation for this class yet.
    """

    def __init__(self):
        Matcher.__init__(self, '\*\s*SOURCE FILE:(\s+)(.+)')

    def apply(self, stack, line):
        m = re.search(self.regex, line)
        #print line
        if len(stack) > 0:
            if m:
                stack[-1].source_file = m.group(2).strip()


class EffectMatcher(Matcher):
    """No documentation for this class yet.
    """

    def __init__(self):
        Matcher.__init__(self, 'EFFECTS NAME IS(\s+)(.+)')

    def apply(self, stack, line):
        m = re.search(self.regex, line)
        if m:
            stack[-1].transition.effect = m.group(2).strip()


class TimewarpMatcher(Matcher):
    """No documentation for this class yet.
    """

    def __init__(self, fps):
        self.fps = fps
        self.regexp = 'M2\s+(\w+)\s+(\-*\d+\.\d+)\s+(\d+:\d+:\d+[\:\;]\d+)'
        #self.regexp = 'M2\s+(\S+)\s+(\S+)\s+(\S+)'
        Matcher.__init__(self, self.regexp)

    def apply(self, stack, line):
        m = re.search(self.regexp, line)
        if m:
            stack[-1].timewarp = \
                Timewarp(m.group(1), m.group(2), m.group(3), self.fps)
            if float(m.group(2)) < 0:
                stack[-1].timewarp.reverse = True


class EventMatcher(Matcher):
    """No documentation for this class yet.
    """

    def __init__(self, fps):
        regexp = re.compile(
            r"(\d+)\s+(\S+)\s+(\S+)\s+(\S+)\s+(\S*)\s+(\d{1,2}:\d{1,2}:\d{1,2}"
            r"[:;]\d{1,3})\s+(\d{1,2}:\d{1,2}:\d{1,2}[:;]\d{1,3})\s+(\d{1,2}:"
            r"\d{1,2}:\d{1,2}[:;]\d{1,3})\s+(\d{1,2}:\d{1,2}:\d{1,2}[:;]"
            r"\d{1,3})")
        Matcher.__init__(self, regexp)
        self.fps = fps

    @classmethod
    def stripper(cls, in_string):
        return in_string.strip()

    def apply(self, stack, line):
        evt = None
        m = re.search(self.regex, line.strip())
        if m:
            matches = m.groups()
            keys = ['num', 'reel', 'track', 'tr_code', 'aux', 'src_start_tc',
                    'src_end_tc', 'rec_start_tc', 'rec_end_tc']
            values = map(self.stripper, matches)
            evt = Event(dict(zip(keys, values)))
            t = evt.tr_code
            if t == 'C':
                if len(stack) > 0:
                    stack[-1].next_event = evt
                evt.transition = Cut()
            elif t == 'D':
                evt.transition = Dissolve()
            elif re.match('W\d+', t):
                evt.transition = Wipe()
            elif t == 'K':
                evt.transition = Key()
            else:
                evt.transition = None
            evt.src_start_tc = pytimecode.PyTimeCode(self.fps,
                                                     evt.src_start_tc)
            evt.src_end_tc = pytimecode.PyTimeCode(self.fps, evt.src_end_tc)
            evt.rec_start_tc = pytimecode.PyTimeCode(self.fps,
                                                     evt.rec_start_tc)
            evt.rec_end_tc = pytimecode.PyTimeCode(self.fps, evt.rec_end_tc)
            stack.append(evt)
        return evt


class Effect(object):
    """No documentation for this class yet.
    """

    def __init__(self):
        pass


class Cut(Effect):
    """No documentation for this class yet.
    """

    def __init__(self):
        Effect.__init__(self)


class Wipe(Effect):
    """No documentation for this class yet.
    """

    def __init__(self):
        Effect.__init__(self)


class Dissolve(Effect):
    """No documentation for this class yet.
    """

    def __init__(self):
        Effect.__init__(self)


class Key(Effect):
    """No documentation for this class yet.
    """

    def __init__(self):
        Effect.__init__(self)


class Timewarp(object):
    """No documentation for this class yet.
    """

    def __init__(self, reel, warp_fps, tc, fps):
        self.reverse = False
        self.reel = reel
        self.fps = fps
        self.warp_fps = float(warp_fps)
        self.timecode = pytimecode.PyTimeCode(fps, tc)

    def to_string(self):
        """the string representation of this Timewarp instance
        """
        return 'M2   %(reel)-8s %(warp_fps)s %(timecode)32s ' % {
            'reel': self.reel,
            'warp_fps': self.warp_fps,
            'timecode': self.timecode
        }


class Event(object):
    """Represents an edit event (or, more specifically, an EDL line denoting a
    clip being part of an EDL event)
    """

    def __init__(self, options):
        """Initialisation function with options:
        """
        self.comments = []
        self.timewarp = None
        self.next_event = None
        self.track = None
        self.clip_name = None
        self.source_file = None
        self.transition = None
        self.aux = None
        self.reel = None
        self.rec_end_tc = None
        self.rec_start_tc = None
        self.src_end_tc = None
        self.src_start_tc = None
        self.num = None
        self.tr_code = None

        # TODO: This is absolutely wrong and not safe. Please validate the
        #       incoming values, before adopting them as instance variables
        #       and instance methods
        for o in options:
            self.__dict__[o] = options[o]

    # def __repr__(self):
    #     v = ["(\n"]
    #     for k in self.__dict__:
    #         v.append("    %s=%s,\n" % (k, self.__dict__[k]))
    #     v.append(")")
    #     return ''.join(v)

    def to_string(self):
        """Human Readable string representation of edl event.

        Returns the string representation of this Event which is suitable
        to be written to a file to gather back the EDL itself.
        """
        effect = ''
        if self.transition:
            try:
                effect = 'EFFECTS NAME IS %s\n' % self.transition.effect
            except AttributeError:
                pass

        s = "%(num)-4s %(reel)-8s %(track)-5s %(tr_code)-4s %(aux)-3s " \
            "%(src_start_tc)s %(src_end_tc)s %(rec_start_tc)s " \
            "%(rec_end_tc)s\n" \
            "%(effect)s" \
            "%(notes)s" \
            "%(timewarp)s" % {
            'num': self.num,
            'reel': self.reel,
            'track': self.track,
            'aux': self.aux,
            'tr_code': self.tr_code,
            'src_start_tc': self.src_start_tc,
            'src_end_tc': self.src_end_tc,
            'rec_start_tc': self.rec_start_tc,
            'rec_end_tc': self.rec_end_tc,
            'effect': effect,
            'notes': '%s\n' % '\n'.join(self.comments)
                if self.comments else '',
            'timewarp': '%s\n' % self.timewarp.to_string()
                if self.has_timewarp() else ''}

        return s

    def to_inspect(self):
        """Human Readable string representation of edl event.
        """
        return self.__repr__()

    def get_comments(self):
        """Return comments array
        """
        return self.comments

    def outgoing_transition_duration(self):
        """TBC
        """
        if self.next_event:
            return self.next_event.incoming_transition_duration()
        else:
            return 0

    def reverse(self):
        """Returns true if clip is timewarp reversed
        """
        return self.timewarp and self.timewarp.reverse

    def copy_properties_to(self, event):
        """Copy event properties to another existing event object
        """
        for k in self.__dict__:
            event.__dict__[k] = self.__dict__[k]
        return event

    def has_transition(self):
        """Returns true if clip if clip uses a transition and not a Cut
        """
        return not isinstance(self.transition, Cut)

    def incoming_transition_duration(self):
        """Returns incoming transition duration in frames, returns 0 if no
        transition set
        """
        d = 0
        if not isinstance(self.transition, Cut):
            d = int(self.aux)
        return d

    def ends_with_transition(self):
        """Returns true if the clip ends with a transition (if the next clip
        starts with a transition)
        """
        if self.next_event:
            return self.next_event.has_transition()
        else:
            return False

    def has_timewarp(self):
        """Returns true if the clip has a timewarp (speed ramp, motion memory,
        you name it)
        """
        if isinstance(self.timewarp, Timewarp):
            return True
        else:
            return False

    def black(self):
        """Returns true if event is black slug
        """
        return self.reel == "BL"

    def rec_length(self):
        """Returns record length of event in frames before transition
        """
        return self.rec_end_tc.frames - self.rec_start_tc.frames

    def rec_length_with_transition(self):
        """Returns record length of event in frames including transition
        """
        return self.rec_length() + self.outgoing_transition_duration()

    def src_length(self):
        """Returns source length of event in frames before transition
        """
        return self.src_end_tc.frames - self.src_start_tc.frames

    def capture_from_tc(self):
        raise NotImplementedError

    def capture_to_and_including_tc(self):
        raise NotImplementedError

    def capture_to_tc(self):
        raise NotImplementedError

    def speed(self):
        raise NotImplementedError

    def generator(self):
        raise NotImplementedError

    def get_clip_name(self):
        return self.clip_name

    def get_reel(self):
        return self.reel

    def event_number(self):
        return self.num

    def get_track(self):
        return self.track

    def get_tr_code(self):
        return self.tr_code

    def get_aux(self):
        return self.aux


class Parser(object):
    """No documentation for this class yet.
    """

    default_fps = 25.0

    def __init__(self, fps):
        if fps is None:
            self.fps = self.default_fps
        self.fps = fps

    def get_matchers(self):
        return [TitleMatcher(), EventMatcher(self.fps), EffectMatcher(),
                NameMatcher(), SourceMatcher(), TimewarpMatcher(self.fps),
                CommentMatcher()]

    def parse(self, input_):
        stack = None
        if isinstance(input_, str):
            input_ = input_.splitlines(True)
        if isinstance(input_, collections.Iterable):
            stack = List(self.fps)
            for l in input_:
                for m in self.get_matchers():
                    m.apply(stack, l)
        pprint.PrettyPrinter(indent=4)
        return stack


edl/edlTest.py
#!/usr/bin/python
"""
	Module EDL
	unit test suite
"""

import edl
import unittest

class TestEDLFunctions(unittest.TestCase):

	def setUp(self):
		pass
		
	def test_24fps(self):
		p=edl.Parser('24')
		with open('test-edls/test_24.edl') as f:
			s=p.parse(f)
			self.assertEqual(s.events[0].clip_name,'clip 1','Failed clip name test')
			self.assertEqual(s.events[0].src_length(),1440,'Wrong source frame length')
			self.assertEqual(s.events[0].rec_length(),1440,'Wrong record frame length')
			self.assertEqual(s.events[0].src_start_tc.frames,86400,'Wrong source start timecode')
			self.assertEqual(s.events[0].src_end_tc.frames,87840,'Wrong source end timecode')
			self.assertEqual(s.events[0].rec_start_tc.frames,0,'Wrong record start timecode')
			self.assertEqual(s.events[0].rec_end_tc.frames,1440,'Wrong record end timecode')
			self.assertEqual(s.events[1].clip_name,'clip #2','Failed clip name test char 2')
			self.assertEqual(s.events[2].clip_name,'clip -3','Failed clip name test char 3')
			self.assertEqual(s.events[3].clip_name,'clip $4','Failed clip name test char 4')
			self.assertEqual(s.events[4].clip_name,'clip &5','Failed clip name test char 5')
			self.assertEqual(s.events[5].src_start_tc.frames,697,"Wrong Source start complex event")
			self.assertEqual(s.events[5].src_end_tc.frames,697,"Wrong Source end complex event")
			self.assertEqual(s.events[5].rec_start_tc.frames,2857,"Wrong Source start complex event")
			self.assertEqual(s.events[5].rec_end_tc.frames,2857,"Wrong Source end complex event")
			
	
	def test_pal(self):
		p=edl.Parser('25')
		with open('test-edls/test_25.edl') as f:
			s=p.parse(f)
	
	def test_ntsc(self):
		p=edl.Parser('29.97')
		with open('test-edls/test_2997NDF.edl') as f:
			s=p.parse(f)

	def test_2398fps(self):
		p=edl.Parser('23.98')
		with open('test-edls/test_2398.edl') as f:
			s=p.parse(f)
			
if __name__== '__main__':
	unittest.main()